<a href="https://colab.research.google.com/github/DLCBA/Chronos/blob/main/Analisis_Sentimientos_Adidas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Instalación de librerías
!pip install transformers torch

  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (14.1 MB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl (731.7 MB)
  Using cached nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl (410.6 MB)
  Using cached nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl (121.6 MB)
  Using cached nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl (56.5 MB)
  Using cached nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl (124.2 MB)
  Using cached nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl (196.0 MB)
  Using cached nvidia_nccl_cu12-2.19.3-py3-none-manylinux1_x86_64.whl (166.0 MB)
  Using cached nvidia_nvtx_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (99 kB)
  Using cached nvidia_nvjitlink_cu12-12.4.127-py3-none-m

In [ ]:
from transformers import pipeline

In [ ]:
# Análisis de sentimientos
sentiment_analyzer = pipeline('sentiment-analysis')

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

In [ ]:
# Texto de entrada
text_sentiment = """
El objetivo principal de este proyecto en Power BI es analizar y visualizar las principales
tendencias de ventas de Adidas durante los años 2020 y 2021. A través de la creación de
dashboards interactivos y gráficos intuitivos, se busca proporcionar una comprensión clara
y detallada de la dinámica de ventas, identificando patrones, fluctuaciones estacionales y
puntos destacados en la actividad comercial.
"""

In [ ]:
# Dividir el texto en oraciones o párrafos
sentiment_segments = text_sentiment.split('.')
results_sentiment = [sentiment_analyzer(segment.strip()) for segment in sentiment_segments if segment.strip()]
print("Análisis de Sentimientos:")
for segment, result in zip(sentiment_segments, results_sentiment):
    if segment.strip():
        print(f"Texto: {segment.strip()}\nSentimiento: {result}\n")

Análisis de Sentimientos:
Texto: El objetivo principal de este proyecto en Power BI es analizar y visualizar las principales
tendencias de ventas de Adidas durante los años 2020 y 2021
Sentimiento: [{'label': 'POSITIVE', 'score': 0.779146134853363}]

Texto: A través de la creación de
dashboards interactivos y gráficos intuitivos, se busca proporcionar una comprensión clara
y detallada de la dinámica de ventas, identificando patrones, fluctuaciones estacionales y
puntos destacados en la actividad comercial
Sentimiento: [{'label': 'NEGATIVE', 'score': 0.984379768371582}]



In [ ]:
# Traducción de texto (de español a inglés)
translator = pipeline('translation', model='Helsinki-NLP/opus-mt-es-en')
text_translation = """
El objetivo principal de este proyecto en Power BI es analizar y visualizar las principales
tendencias de ventas de Adidas durante los años 2020 y 2021. A través de la creación de
dashboards interactivos y gráficos intuitivos, se busca proporcionar una comprensión clara
y detallada de la dinámica de ventas, identificando patrones, fluctuaciones estacionales y
puntos destacados en la actividad comercial.
"""
result_translation = translator(text_translation)
print("\nTraducción de Texto:")
print(result_translation)

config.json:   0%|          | 0.00/1.44k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/312M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/293 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/44.0 [00:00<?, ?B/s]

source.spm:   0%|          | 0.00/826k [00:00<?, ?B/s]

target.spm:   0%|          | 0.00/802k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.59M [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/models/marian/tokenization_marian.py:175: UserWarning: Recommended: pip install sacremoses.
  warnings.warn("Recommended: pip install sacremoses.")



Traducción de Texto:
[{'translation_text': 'The main objective of this project in Power BI is to analyze and visualize the main sales trends of Adidas during the years 2020 and 2021. Through the creation of interactive dashboards and intuitive graphics, it seeks to provide a clear and detailed understanding of the dynamics of sales, identifying patterns, seasonal fluctuations and highlights in commercial activity.'}]


In [ ]:
# Resumen de texto
summarizer = pipeline('summarization')
text_summary = """
El objetivo principal de este proyecto en Power BI es analizar y visualizar las principales
tendencias de ventas de Adidas durante los años 2020 y 2021. A través de la creación de
dashboards interactivos y gráficos intuitivos, se busca proporcionar una comprensión clara
y detallada de la dinámica de ventas, identificando patrones, fluctuaciones estacionales y
puntos destacados en la actividad comercial.
"""
result_summary = summarizer(text_summary, max_length=50, min_length=25, do_sample=False)
print("\nResumen de Texto:")
print(result_summary)

No model was supplied, defaulted to sshleifer/distilbart-cnn-12-6 and revision a4f8f3e (https://huggingface.co/sshleifer/distilbart-cnn-12-6).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/1.80k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.22G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]


Resumen de Texto:
[{'summary_text': ' Power BI is analizar and visualizar las principales de ventas de Adidas de 2020 and 2021 . A través de la creación, Power BI aims to proporcionar una comprencia'}]


Mejoramiento de los resultados

In [ ]:
# Usar un modelo preentrenado específico para el análisis de sentimientos en español
sentiment_analyzer = pipeline('sentiment-analysis', model='nlptown/bert-base-multilingual-uncased-sentiment')

config.json:   0%|          | 0.00/953 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/669M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/39.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/872k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

In [ ]:
# Texto de entrada
text_sentiment = """
El objetivo principal de este proyecto en Power BI es analizar y visualizar las principales
tendencias de ventas de Adidas durante los años 2020 y 2021. A través de la creación de
dashboards interactivos y gráficos intuitivos, se busca proporcionar una comprensión clara
y detallada de la dinámica de ventas, identificando patrones, fluctuaciones estacionales y
puntos destacados en la actividad comercial.
"""

In [ ]:
# Dividir el texto en oraciones o párrafos
import nltk
nltk.download('punkt')
from nltk.tokenize import sent_tokenize

sentiment_segments = sent_tokenize(text_sentiment, language='spanish')
results_sentiment = [sentiment_analyzer(segment) for segment in sentiment_segments if segment.strip()]


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


In [ ]:
print("Análisis de Sentimientos:")
for segment, result in zip(sentiment_segments, results_sentiment):
    if segment.strip():
        print(f"Texto: {segment.strip()}\nSentimiento: {result}\n")

Análisis de Sentimientos:
Texto: El objetivo principal de este proyecto en Power BI es analizar y visualizar las principales
tendencias de ventas de Adidas durante los años 2020 y 2021.
Sentimiento: [{'label': '4 stars', 'score': 0.4642845690250397}]

Texto: A través de la creación de
dashboards interactivos y gráficos intuitivos, se busca proporcionar una comprensión clara
y detallada de la dinámica de ventas, identificando patrones, fluctuaciones estacionales y
puntos destacados en la actividad comercial.
Sentimiento: [{'label': '4 stars', 'score': 0.5384299159049988}]



**Explicación de la salida**

**Texto 1:**

Texto: "El objetivo principal de este proyecto en Power BI es analizar y visualizar las principales tendencias de ventas de Adidas durante los años 2020 y 2021."
Sentimiento: [{'label': '4 stars', 'score': 0.4642845690250397}]
Interpretación: Este texto tiene una clasificación de 4 estrellas con un puntaje de confianza de aproximadamente 0.46. Esto sugiere que el sentimiento asociado con este texto es positivo, pero no extremadamente fuerte. La descripción del proyecto parece ser vista favorablemente, pero con cierta moderación.

**Texto 2:**

Texto: "A través de la creación de dashboards interactivos y gráficos intuitivos, se busca proporcionar una comprensión clara y detallada de la dinámica de ventas, identificando patrones, fluctuaciones estacionales y puntos destacados en la actividad comercial."
Sentimiento: [{'label': '4 stars', 'score': 0.5384299159049988}]
Interpretación: Este texto también recibe una clasificación de 4 estrellas, con un puntaje de confianza de aproximadamente 0.54. Similar al primer texto, esto indica un sentimiento positivo, con una mayor confianza en la evaluación positiva. La explicación de las herramientas y objetivos del proyecto es percibida de manera positiva.

Ajustando el código

In [ ]:
# Usar el modelo de análisis de sentimientos específico para español
sentiment_analyzer = pipeline('sentiment-analysis', model='pysentimiento/robertuito-sentiment-analysis')

# Textos de entrada
texts = [
    "El objetivo principal de este proyecto en Power BI es analizar y visualizar las principales tendencias de ventas de Adidas durante los años 2020 y 2021.",
    "A través de la creación de dashboards interactivos y gráficos intuitivos, se busca proporcionar una comprensión clara y detallada de la dinámica de ventas, identificando patrones, fluctuaciones estacionales y puntos destacados en la actividad comercial."
]

# Análisis de Sentimientos
results_sentiment = [sentiment_analyzer(text) for text in texts]

print("Análisis de Sentimientos:")
for text, result in zip(texts, results_sentiment):
    print(f"Texto: {text}\nSentimiento: {result}\n")


config.json:   0%|          | 0.00/925 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/435M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/384 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.31M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/167 [00:00<?, ?B/s]

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


Análisis de Sentimientos:
Texto: El objetivo principal de este proyecto en Power BI es analizar y visualizar las principales tendencias de ventas de Adidas durante los años 2020 y 2021.
Sentimiento: [{'label': 'NEU', 'score': 0.5346047282218933}]

Texto: A través de la creación de dashboards interactivos y gráficos intuitivos, se busca proporcionar una comprensión clara y detallada de la dinámica de ventas, identificando patrones, fluctuaciones estacionales y puntos destacados en la actividad comercial.
Sentimiento: [{'label': 'NEU', 'score': 0.5920534133911133}]



**Texto 1:**

Texto: "El objetivo principal de este proyecto en Power BI es analizar y visualizar las principales tendencias de ventas de Adidas durante los años 2020 y 2021."
Sentimiento: [{'label': 'NEU', 'score': 0.5346047282218933}]
Interpretación: El análisis de sentimientos ha clasificado este texto como neutral con una puntuación de confianza de aproximadamente 0.53. Esto sugiere que el texto no expresa emociones fuertes ni positivas ni negativas. El contenido del texto es informativo y describe el objetivo del proyecto de una manera objetiva, lo que es consistente con una evaluación neutral.

**Texto 2:**

Texto: "A través de la creación de dashboards interactivos y gráficos intuitivos, se busca proporcionar una comprensión clara y detallada de la dinámica de ventas, identificando patrones, fluctuaciones estacionales y puntos destacados en la actividad comercial."
Sentimiento: [{'label': 'NEU', 'score': 0.5920534133911133}]
Interpretación: Este texto también ha sido clasificado como neutral con una puntuación de confianza de aproximadamente 0.59. Similar al primer texto, se trata de una descripción objetiva de las funcionalidades y objetivos del proyecto, sin expresiones de emociones positivas o negativas.

**Los resultados indican que los textos proporcionados son objetivos y no evocan emociones fuertes, lo que es coherente con la naturaleza descriptiva e informativa de los textos.**

In [ ]:
# Crear un pipeline para resumir texto
summarizer = pipeline("summarization", model="t5-base", tokenizer="t5-base", framework="tf")

# Texto de entrada
text = """
El objetivo principal de este proyecto en Power BI es analizar y visualizar las principales tendencias de ventas de Adidas durante los años 2020 y 2021. A través de la creación de dashboards interactivos y gráficos intuitivos, se busca proporcionar una comprensión clara y detallada de la dinámica de ventas, identificando patrones, fluctuaciones estacionales y puntos destacados en la actividad comercial.
"""

# Generar un resumen del texto
summary = summarizer(text, max_length=150, min_length=30, do_sample=False)

# Imprimir el resumen
print("Resumen de Texto:")
print(summary[0]['summary_text'])


config.json:   0%|          | 0.00/1.21k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/892M [00:00<?, ?B/s]

All PyTorch model weights were used when initializing TFT5ForConditionalGeneration.

All the weights of TFT5ForConditionalGeneration were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFT5ForConditionalGeneration for predictions without further training.


spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.39M [00:00<?, ?B/s]

Resumen de Texto:
el objetivo principal de este proyecto en Power BI es analizar y visualizar las principales tendencias de ventas de Adidas durante los aos 2020 y 2021.


Otras alternativas a evaluar

Reconocimiento de entidades nombradas (NER)

In [ ]:
# Crear un pipeline para NER (Reconocimiento de Entidades Nombradas)
ner_pipeline = pipeline("ner", framework="tf", model="dbmdz/bert-large-cased-finetuned-conll03-english")

# Texto de entrada
text = """
El objetivo principal de este proyecto en Power BI es analizar y visualizar las principales tendencias de ventas de Adidas durante los años 2020 y 2021. A través de la creación de dashboards interactivos y gráficos intuitivos, se busca proporcionar una comprensión clara y detallada de la dinámica de ventas, identificando patrones, fluctuaciones estacionales y puntos destacados en la actividad comercial.
"""

# Realizar el reconocimiento de entidades nombradas
ner_results = ner_pipeline(text)

# Imprimir los resultados
print("Entidades Nombradas encontradas en el texto:")
for entity in ner_results:
    print(f"Entidad: {entity['word']}, Etiqueta: {entity['entity']}")


config.json:   0%|          | 0.00/998 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.33G [00:00<?, ?B/s]

All PyTorch model weights were used when initializing TFBertForTokenClassification.

All the weights of TFBertForTokenClassification were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFBertForTokenClassification for predictions without further training.


tokenizer_config.json:   0%|          | 0.00/60.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

Entidades Nombradas encontradas en el texto:
Entidad: Power, Etiqueta: I-ORG
Entidad: B, Etiqueta: I-ORG
Entidad: ##I, Etiqueta: I-ORG
Entidad: Ad, Etiqueta: I-ORG
Entidad: ##idas, Etiqueta: I-ORG


**Explicación de la salida**

Entidad: Power, Etiqueta: I-ORG: La entidad "Power" ha sido identificada como parte de una organización (ORG). La etiqueta "I-ORG" indica que esta es una entidad de organización y que es una parte continua de una entidad más larga.

Entidad: B, Etiqueta: I-ORG: La entidad "B" también ha sido identificada como parte de una organización (ORG). Al igual que la anterior, la etiqueta "I-ORG" indica que esta es una entidad de organización y que es una parte continua de una entidad más larga.

Entidad: ##I, Etiqueta: I-ORG: Esta entidad parece ser parte de una organización, pero la forma en que está tokenizada indica que el modelo considera que "I" es una parte de la entidad y "##I" puede ser un prefijo o una parte de una palabra que sigue a "I".

Entidad: Ad, Etiqueta: I-ORG: Similar a las anteriores, "Ad" ha sido identificado como parte de una organización (ORG).

Entidad: ##idas, Etiqueta: I-ORG: Al igual que "##I", "##idas" parece ser una parte de una palabra o un sufijo que sigue a una entidad de organización.

En resumen, el modelo ha identificado varias entidades que considera partes de una organización (ORG) en el texto. La presencia de prefijos como "##" indica que algunas de estas entidades pueden ser partes de palabras que siguen a la entidad principal.








**Generación de Resúmenes Abstractivos**

In [ ]:
from transformers import T5ForConditionalGeneration, T5Tokenizer

# Cargar el modelo y el tokenizador T5
model_name = "t5-small"
tokenizer = T5Tokenizer.from_pretrained(model_name)
model = T5ForConditionalGeneration.from_pretrained(model_name)

# Texto de entrada
text = """
El objetivo principal de este proyecto en Power BI es analizar y visualizar las principales tendencias de ventas de Adidas durante los años 2020 y 2021. A través de la creación de dashboards interactivos y gráficos intuitivos, se busca proporcionar una comprensión clara y detallada de la dinámica de ventas, identificando patrones, fluctuaciones estacionales y puntos destacados en la actividad comercial.
"""

# Tokenizar el texto de entrada
inputs = tokenizer.encode("summarize: " + text, return_tensors="pt", max_length=512, truncation=True)

# Generar el resumen
summary_ids = model.generate(inputs, max_length=150, min_length=40, length_penalty=2.0, num_beams=4, early_stopping=True)

# Decodificar y mostrar el resumen
summary = tokenizer.decode(summary_ids[0], skip_special_tokens=True)
print("Resumen del texto:")
print(summary)


tokenizer_config.json:   0%|          | 0.00/2.32k [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.39M [00:00<?, ?B/s]

You are using the default legacy behaviour of the <class 'transformers.models.t5.tokenization_t5.T5Tokenizer'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


config.json:   0%|          | 0.00/1.21k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/242M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/147 [00:00<?, ?B/s]

Resumen del texto:
objetivo principal de este proyecto en Power BI es analizar y visualizar las principales tendencias de ventas de Adidas durante los aos 2020 y 2021.


La salida del resumen es similar a la anterior, mejorando la realizada por primera vez.

**Conclusiones**

**El procesamiento de lenguaje natural (PLN**) es un campo de la inteligencia artificial que se enfoca en la interacción entre las computadoras y el lenguaje humano.
A continuación se explican las mejoras realizadas en tres tareas de PLN: análisis de sentimientos, generación de resúmenes y reconocimiento de entidades nombradas (NER). Se describe el enfoque original y las mejoras implementadas en cada tarea para optimizar la precisión y la relevancia de los resultados.

**Análisis de Sentimientos:**

**Enfoque Original:** Utilización de un modelo preentrenado para el análisis de sentimientos con un enfoque general.

**Mejora:** Cambio al uso de un modelo específico para español, como nlptown/bert-base-multilingual-uncased-sentiment, para mejorar la precisión del análisis de sentimientos en textos en español.

**Generación de Resúmenes:**

**Enfoque Original**: Utilización de un modelo preentrenado para la generación de resúmenes con parámetros predeterminados.

**Mejora:** Ajuste de los parámetros del modelo T5, como max_length, min_length, num_beams, etc., para mejorar la calidad y la relevancia del resumen generado.

**Reconocimiento de Entidades Nombradas (NER):**

**Enfoque Original:** Utilización de un modelo preentrenado para NER con un enfoque general.
**Mejora:** Empleo de un modelo específico para reconocimiento de entidades nombradas en inglés, como dbmdz/bert-large-cased-finetuned-conll03-english, para obtener resultados más precisos en textos en inglés.

Las mejoras implementadas en las tareas de PLN han permitido optimizar la precisión y la relevancia de los resultados obtenidos. La selección de modelos específicos para cada tarea y la configuración adecuada de los parámetros han contribuido significativamente a mejorar el rendimiento del sistema en términos de calidad de salida. Estas mejoras facilitan una mejor comprensión y aplicación de las técnicas de PLN en diversos contextos y dominios de aplicación.

Este informe destaca la importancia de seleccionar modelos apropiados y ajustar los parámetros para lograr resultados óptimos en las tareas de procesamiento de lenguaje natural.








